src/server.c

```diff
@@ -163,25 +167,40 @@ void _serverLog(int level, const char *fmt, ...) {
  * of view of Redis. Signals that are going to kill the server anyway and
  * where we need printf-alike features are served by serverLog(). */
 void serverLogFromHandler(int level, const char *msg) {
-    int fd;
-    int log_to_stdout = server.logfile[0] == '\0';
     char buf[64];
 
-    if ((level&0xff) < server.verbosity || (log_to_stdout && server.daemonize))
+    if ((level&0xff) < server.verbosity || (server.log_fd == STDOUT_FILENO && server.daemonize))
         return;
-    fd = log_to_stdout ? STDOUT_FILENO :
-                         open(server.logfile, O_APPEND|O_CREAT|O_WRONLY, 0644);
-    if (fd == -1) return;
+    if (server.log_fd == -1) return;
     ll2string(buf,sizeof(buf),getpid());
-    if (write(fd,buf,strlen(buf)) == -1) goto err;
-    if (write(fd,":signal-handler (",17) == -1) goto err;
+    if (write(server.log_fd,buf,strlen(buf)) == -1) return;
+    if (write(server.log_fd,":signal-handler (",17) == -1) return;
     ll2string(buf,sizeof(buf),time(NULL));
-    if (write(fd,buf,strlen(buf)) == -1) goto err;
-    if (write(fd,") ",2) == -1) goto err;
-    if (write(fd,msg,strlen(msg)) == -1) goto err;
-    if (write(fd,"\n",1) == -1) goto err;
-err:
-    if (!log_to_stdout) close(fd);
+    if (write(server.log_fd,buf,strlen(buf)) == -1) return;
+    if (write(server.log_fd,") ",2) == -1) return;
+    if (write(server.log_fd,msg,strlen(msg)) == -1) return;
+    if (write(server.log_fd,"\n",1) == -1) return;
+}
+
+/* Close the currently open logfile and re-open it based on server.logfile */
+int serverLogOpen(void) {
+    if (server.log_fd != -1 && server.log_fd != STDOUT_FILENO) close(server.log_fd);
+    if (server.logfile[0] == '\0') {
+        server.log_fd = STDOUT_FILENO;
+    } else {
+        server.log_fd = open(server.logfile, O_APPEND|O_CREAT|O_WRONLY|O_CLOEXEC, 0644);
+    }
+    return server.log_fd == -1 ? C_ERR : C_OK;
+}
+
+/* Re-open log file on SIGHUP */
+static void serverLogReopenSigHandler(int signum) {
+    UNUSED(signum);
+
+    if (server.log_fd == -1 || server.log_fd == STDOUT_FILENO) return;
```

Comment:
We might want to at least try to open the logfile again, if someone's disk get's filled up and it fails, they might fix it but we'll never try to rotate it again.

Reply:
You're right!